Add Witch docs (#632)
metastatedev/juvix: doc/Witch/intro-to-witch.md

> ---
> title: Introduction to Witch
> ---

Reviewer: I think we normally do # as the title? Do we want to do that here as well for consistency?
```suggestion
# Introduction to Witch
```

Author: Sorry, my env auto-gen markdown, and that's usually the format static generators use.

To popularise software verification, we must make proof writing less of
a burden on you, the user. Currently, only specialists and a few others 
can write proofs for their programs, even though you surely understand
the domain and invariants of your projects -- otherwise you wouldn't
be able to write any software at all.

> *Our hypothesis is that you, programmers, are

Reviewer 2: ```suggestion
*Our hypothesis is that programmers, are
```

Author: I'd prefer to keep "you" rather than "programmers"? I think it should sound like a direct conversation than some sort of impersonal technical piece. Idk, I think it feels better.

Author: @thealmarty what do you think?

Reviewer: I think keeping the "you" is fine if that's what you prefer. It matches the style with the rest of the docs.

well-equipped to derive and prove properties of the software they write,
but they lack the mathematical maturity and vocabulary to carry out a
formal proof.* We have evidence that students fail to produce well-made
proofs due to the lack of mathematical maturity, even though they do
understand the subject matter at hand.

> We believe that program verification don't have to be like this,

Reviewer: ```suggestion
We believe that program verification doesn't have to be like this,
```

and we did something about it. We created an assistant tool for theorem provers
that puts together many different strategies for proof automation. Our
specification for such an assistant for Juvix uses algebraic effects and
handlers as the means of congregation.

Our approach is named Witch, a play on the
assistant tools colloquially called "wizards". There is no consensus of what a

> wizard is or what exactly the tasks are it is supposed assist with. Wizards seem

Reviewer: ```suggestion
wizard is or what the exact tasks it is supposed to assist with. Wizards seem
```
seems like an awkward sentence here, hope this helps.

to be used mostly for multiple-step and/or configuration features, however. We
went for the name "witch" to align it to the idea of assistant tools,
while dodging the overloaded and confusing terminology.

We propose an effects and handlers view of such proofs,
based on prior work developed on the [Andromeda
prover](https://www.andromeda-prover.org).  
In Juvix, you program as you would normally and
invoke the proof environment as an effect to prove certain properties as
you go. Given that the proof environment is *just* an effect, different proof

> styles (e.g., Agda-style dependent types , 

Reviewer: ```suggestion
styles (e.g., Agda-style dependent types, 
```

1.  as proofs cannot be fully automated, all approaches that try to
    automate the process (e.g, proof search, SMT solver) may be
    non-deterministic or never find a solution. Therefore, the system
    should be able to handle "impure" computations and errors. Algebraic
    effects and handlers have well-defined semantics and provide a
    simple interface for performing effects. With them, we avoid
    indiscriminate effects that are often error-prone and intricate
    effects machinery such as monad transformers;

2.  the semantics of effects and handlers accommodate composition of
    arbitrary effects and the composition of multiple handlers, which
    means users have the ability to weaken more general strategies into
    specific ones while maintaining top-level handlers unmodified;

3.  effects and handlers have well-defined semantics, and it is the main
    feature of many new languages, such as
    [Koka](https://koka-lang.github.io/koka/doc/index.html),
    [Eff](https://www.eff-lang.org), [Frank](https://github.com/frank-lang) and 
    [Effekt](https://effekt-lang.org), and have implementation in all major

>     functional programming languages, what guarantees that our approach is based

Reviewer: ```suggestion
    functional programming languages. This guarantees that our approach is based
```

    in research and real-world experiences, and not something we hacked together
    over the weekend. 

## The Essence of Witch

As for the syntax, we use `operation { params } via handler`, which is

> a syntactic improvement over `handler(operation, params)`, since effect handling

Reviewer 2: ```suggestion
semantically equivalent to but a syntactic improvement over `handler(operation, params)`, since effect handling
```

is similar to function application, but also carries effect information.
The user defines data types and functions as usual, and then uses
Witch to prove properties about said definitions. The Examples below
shows it in action.

> ![](paper/image/witch.svg)

Reviewer: This didn't show in the rendered version?

Author: Err, forgot to add when moved it away from witch's repo. Thanks for catching it.

The proof of commutativity of addition under
natural numbers and of associativity of list concatenation are shown,
and use the three main effects: Solver, Proof Search and Tactics. In the
proof assistant literature, there exists no precise definition of
commonly used terms "solver", "proof search" and "tactics". All these
terms are used in different communities, and mean some variation of
"automatic strategy to construct a term under certain constraints".

In Witch, however, we use the following definitions of previously known terms:

-   The Solver effect is used for integration with external solvers via

>     IO; we believe should suffice for the user to write, e.g. `SMT`, and

Reviewer 2: ```suggestion
    IO; we believe it should suffice for the user to write, e.g. `SMT`, and
```

    handlers should implement internal strategies to choose between the
    different theories supported by solvers. If the black-box approach
    to solvers presents itself a challenge, a specialisation of the
    handler is possible, e.g. `operation { params } via Z3.QF-UFLIA`
    .

-   The Proof Search effect is used for library-level algorithms; users
    may choose to implement their own algorithms using a limited set of

>     meta-programming[^1] constructs that are handled at top-level[^2].

Reviewer: ```suggestion
    meta-programming<sup>1</sup> constructs that are handled at top-level<sup>2</sup>.
```
Maybe making these superscripts will look better?

Author: That's how GitHub-flavoured markdown does footnotes, we can change what's within brackets, but there's nothing else we can do.
